src/Razor/Microsoft.CodeAnalysis.Razor/src/DefaultTagHelperDescriptorProvider.cs

```diff
@@ -34,17 +34,23 @@ public void Execute(TagHelperDescriptorProviderContext context)
 
             var types = new List<INamedTypeSymbol>();
             var visitor = new TagHelperTypeVisitor(iTagHelper, types);
+            var discoveryMode = context.Items.GetTagHelperDiscoveryFilter();
 
-            // We always visit the global namespace.
-            visitor.Visit(compilation.Assembly.GlobalNamespace);
+            if ((discoveryMode & TagHelperDiscoveryFilter.CurrentCompilation) == TagHelperDiscoveryFilter.CurrentCompilation)
+            {
+                visitor.Visit(compilation.Assembly.GlobalNamespace);
+            }
 
-            foreach (var reference in compilation.References)
+            if ((discoveryMode & TagHelperDiscoveryFilter.ReferenceAssemblies) == TagHelperDiscoveryFilter.ReferenceAssemblies)
```

Comment:
Can we encapsulate these types of checks into two static methods: `ShouldDiscoverFromReferenceAssemblies` and `ShouldDiscoverFromAppAssembly` or something?

Reply:
If I were to create a method, it's tempting to also put the lookup inside the code which now results in an extra dictionary lookup per call. In the absence of it, it's a simple enum check, which seems like an overkill to have a method for.